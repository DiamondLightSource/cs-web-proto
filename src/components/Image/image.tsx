import React, { CSSProperties } from "react";
import PropTypes from "prop-types";

import {
  Component,
  Widget,
  WidgetPropType,
  InferWidgetProps
} from "../Widget/widget";

const ImageProps = {
  src: PropTypes.string.isRequired,
  alt: PropTypes.string,
  fill: PropTypes.bool
};

export const ImageComponent = (
  props: PropTypes.InferProps<typeof ImageProps> & Component
): JSX.Element => {
  let imageSize = undefined;
  let overflow = "auto";
  if (props.fill === true) {
    imageSize = "100%";
    overflow = "hidden";
  }

  const style: CSSProperties = {
    overflow: overflow,
    textAlign: "left",
    ...props.style
  };

  return (
    <div style={style}>
      <img
        src={props.src || undefined}
        alt={props.alt || undefined}
        style={{
          height: imageSize,
          width: imageSize
        }}
      />
    </div>
  );
};

const ImageWidgetProps = {
  ...ImageProps,
  ...WidgetPropType
};

export const Image = (
  props: InferWidgetProps<typeof ImageWidgetProps>
): JSX.Element => <Widget baseWidget={ImageComponent} {...props} />;

Image.propTypes = ImageWidgetProps;
