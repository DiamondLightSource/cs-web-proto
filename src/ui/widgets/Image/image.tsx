import React, { CSSProperties, useContext } from "react";

import { useCommonCss, Widget } from "../widget";
import { WidgetPropType } from "../widgetProps";
import {
  InferWidgetProps,
  StringProp,
  BoolPropOpt,
  StringPropOpt,
  FloatPropOpt,
  FuncPropOpt
} from "../propTypes";
import { registerWidget } from "../register";
import { BaseUrlContext } from "../../../baseUrl";

const ImageProps = {
  imageFile: StringProp,
  alt: StringPropOpt,
  stretchToFit: BoolPropOpt,
  fitToWidth: BoolPropOpt,
  fitToHeight: BoolPropOpt,
  rotation: FloatPropOpt,
  flipHorizontal: BoolPropOpt,
  flipVertical: BoolPropOpt,
  onClick: FuncPropOpt
};

export const ImageComponent = (
  props: InferWidgetProps<typeof ImageProps>
): JSX.Element => {
  const { rotation = 0, flipHorizontal, flipVertical } = props;

  const onClick = (event: React.MouseEvent<HTMLDivElement>): void => {
    if (props.onClick) {
      props.onClick(event);
    }
  };

  const baseUrl = useContext(BaseUrlContext);
  let file = `img/${props.imageFile}`;
  if (!file.startsWith("http")) {
    file = `${baseUrl}/${file}`;
  }
  let imageHeight: string | undefined = undefined;
  let imageWidth: string | undefined = undefined;
  const overflow = "hidden";
  if (props.stretchToFit) {
    imageWidth = "100%";
    imageHeight = "100%";
  } else if (props.fitToWidth) {
    imageWidth = "100%";
  } else if (props.fitToHeight) {
    imageHeight = "100%";
  }

  const style: CSSProperties = {
    ...useCommonCss(props as any),
    overflow,
    textAlign: "left",
    width: imageWidth,
    height: imageHeight
  };

  return (
    <div style={style} onClick={onClick}>
      <img
        src={file}
        alt={props.alt || undefined}
        style={{
          width: imageWidth,
          height: imageHeight,
          display: "block",
          transform: `rotate(${rotation}deg) scaleX(${
            flipHorizontal ? -1 : 1
          }) scaleY(${flipVertical ? -1 : 1})`
        }}
      />
    </div>
  );
};

const ImageWidgetProps = {
  ...ImageProps,
  ...WidgetPropType
};

export const Image = (
  props: InferWidgetProps<typeof ImageWidgetProps>
): JSX.Element => <Widget baseWidget={ImageComponent} {...props} />;

registerWidget(Image, ImageWidgetProps, "image");
