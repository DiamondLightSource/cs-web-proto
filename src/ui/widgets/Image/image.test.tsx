import React from "react";
import { ImageComponent } from "./image";
import { render, screen } from "@testing-library/react";

describe("<ImageComponent />", (): void => {
  describe("Visible tests", (): void => {
    test("it contains an image element", (): void => {
      render(<ImageComponent imageFile="test.png" />);
      expect(screen.getByRole("img")).toBeInTheDocument();

      expect(screen.getByRole("img")).toHaveAttribute("src", "test.png");
    });

    test("it passes alternative text through", (): void => {
      render(<ImageComponent imageFile="test.png" alt="test text" />);

      expect(screen.getByAltText("test text")).toBeInTheDocument();
    });
  });

  describe("prop testing", (): void => {
    test("its source is passed through properly", (): void => {
      render(<ImageComponent imageFile="test.jpg" />);
      expect(screen.getByRole("img")).toHaveAttribute("src", "test.jpg");
    });

    test("flips and rotations are applied", (): void => {
      const imageProps = {
        imageFile: "test.svg",
        flipHorizontal: true,
        flipVertical: true,
        rotation: 45
      };

      render(<ImageComponent {...imageProps} />);
      const img = screen.getByRole("img");

      if ("style" in img) {
        expect(img.style).toHaveProperty(
          "transform",
          "rotate(45deg) scaleX(-1) scaleY(-1)"
        );
      }
      expect.assertions(1);
    });
  });
});
