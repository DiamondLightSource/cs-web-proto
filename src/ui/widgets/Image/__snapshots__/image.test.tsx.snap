// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Image /> it matches the snapshot 1`] = `
<div
  style={
    Object {
      "overflow": "auto",
      "textAlign": "left",
    }
  }
>
  <img
    src="myimage"
    style={
      Object {
        "height": undefined,
        "width": undefined,
      }
    }
  />
</div>
`;
